File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        res = [0]
        def maxH(root):
            if root == None:
                return 0
            
            lh = maxH(root.left)
            rh = maxH(root.right)

            res[0] = max(res[0], lh + rh)

            return 1 + max(lh, rh)
        maxH(root)
        return res[0]

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {        
    int maxVal = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        int test = inorder(root);
        return maxVal;
    }

    int inorder(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int left = inorder(root.left);
        int right = inorder(root.right);
        maxVal = Math.max(maxVal, left + right);
        return 1 + Math.max(left, right);
    }
}